File: 3b.pl
#!/usr/bin/perl
use warnings;
use strict;
use feature qw{ say };

use ARGV::OrDATA;

chomp( my @list = <> );
my @oxygen = my @co2 = @list;

my $pos = 0;
while (@oxygen > 1) {
    my $count1 = grep substr($_, $pos, 1), @oxygen;
    @oxygen = grep substr($_, $pos, 1) == ($count1 >= @oxygen / 2), @oxygen;
    ++$pos;
}
$pos = 0;
while (@co2 > 1) {
    my $count0 = grep ! substr($_, $pos, 1), @co2;
    @co2 = grep substr($_, $pos, 1) != ($count0 <= @co2 / 2), @co2;
    ++$pos;
}
say oct("0b$oxygen[0]") * oct("0b$co2[0]");

__DATA__
00100
11110
10110
10111
10101
01111
00111
11100
10000
11001
00010
01010

File: 3b.py
#! /usr/bin/python3
import sys

def process (binaries, gt):
    pos = 0
    while len(binaries) > 1:
        count1 = sum([b[pos:pos+1] == "1" for b in binaries])
        l2 = len(binaries) / 2

        compare = count1 >= l2 if gt else count1 <= l2
        compare = "1" if compare else "0"

        binaries = list(filter(lambda b: compare == b[pos:pos+1], binaries))
        pos += 1
    return binaries[0]

with open(sys.argv[1], 'r') as fh:
    oxygens = [line[0:-1] for line in fh]

neg  = str.maketrans("01", "10")
co2s = [o.translate(neg) for o in oxygens]

oxygen = process(oxygens, 1)
co2    = process(co2s,    0).translate(neg)
print(int(oxygen, 2) * int(co2, 2))
